src/Core/MailTemplate/MailLayoutFolderCatalog.php

```diff
@@ -0,0 +1,246 @@
+
+    /** @var string */
+    private $mailThemesFolder;
+
+    /** @var HookDispatcherInterface */
+    private $hookDispatcher;
+
+    /**
+     * @param string $mailThemesFolder
+     * @param HookDispatcherInterface $hookDispatcher
+     */
+    public function __construct($mailThemesFolder, HookDispatcherInterface $hookDispatcher)
+    {
+        $this->mailThemesFolder = $mailThemesFolder;
+        $this->hookDispatcher = $hookDispatcher;
+    }
+
+    /**
+     * Returns the list of found themes (non empty folders, in the mail themes
+     * folder).
+     *
+     * @throws FileNotFoundException
+     *
+     * @return MailThemeCollectionInterface
+     */
+    public function listThemes()
+    {
+        $this->checkThemesFolder();
+
+        $finder = new Finder();
+        $finder->directories()->in($this->mailThemesFolder)->depth(0);
+        $mailThemes = new MailThemeCollection();
+        /** @var SplFileInfo $mailThemeFolder */
+        foreach ($finder as $mailThemeFolder) {
+            $dirFinder = new Finder();
+            $dirFinder->files()->in($mailThemeFolder->getRealPath());
+            if ($dirFinder->count() > 0) {
+                $mailThemes[] = new MailTheme($mailThemeFolder->getFilename());
+            }
+        }
+
+        //This hook allows you to add/remove a mail theme
+        $this->hookDispatcher->dispatchWithParameters(
+            MailLayoutCatalogInterface::LIST_MAIL_THEMES_HOOK,
+            ['mailThemes' => $mailThemes]
+        );
+
+        return $mailThemes;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function listLayouts($mailTheme)
+    {
+        $mailThemeFolder = implode(DIRECTORY_SEPARATOR, [$this->mailThemesFolder, $mailTheme]);
+        //This hook allows to change the mail them folder
+        $this->hookDispatcher->dispatchWithParameters(
+            static::GET_MAIL_THEME_FOLDER_HOOK,
+            [
+                'mailTheme' => $mailTheme,
+                'mailThemeFolder' => &$mailThemeFolder,
+            ]
+        );
+        $this->checkThemeFolder($mailThemeFolder);
+
+        $mailThemeLayouts = new MailLayoutCollection();
+        $this->listCoreLayouts($mailThemeLayouts, $mailThemeFolder);
+        $this->listModulesLayouts($mailThemeLayouts, $mailThemeFolder);
+
+        //This hook allows to add/remove layouts to a mail theme
+        $this->hookDispatcher->dispatchWithParameters(
+            MailLayoutCatalogInterface::LIST_MAIL_THEME_LAYOUTS_HOOK,
+            [
+                'mailTheme' => $mailTheme,
+                'mailThemeLayouts' => $mailThemeLayouts,
+            ]
+        );
+
+        return $mailThemeLayouts;
+    }
+
+    /**
+     * @param MailLayoutCollectionInterface $collection
+     * @param string $mailThemeFolder
+     */
+    private function listCoreLayouts(MailLayoutCollectionInterface $collection, $mailThemeFolder)
+    {
+        $coreLayoutsFolder = implode(DIRECTORY_SEPARATOR, [
```

Comment:
same code as line 149

Reply:
no, one is for core folder, the other one for module folder (the last parameter is different `MailTemplateInterface::CORE_CATEGORY` != MailTemplateInterface::MODULES_CATEGORY)
do you think it's worth it to create a function to factorize these two pieces of code?